fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -3,204 +3,683 @@
+        #   * num_inputs and num_outputs for modules matche list of connections defined in the graph.
+        #   * all inputs to a module should come from model input, or modules with smaller index in
+        #     the graph. This condition is used in implementaion of DistributedPipeline.forward. Even
+        #     if we relax this condition, still need to make sure the graph is acyclic.
+        pass
+
+    def compute_output_users(self) -> None:
+        """Precomputs self.model_input_users and self.output_users for internal use by the pipleine
+        class. These two lists show consumers of inputs to the model, and outputs of each module of
+        the graph. Each consumer is a pair (i, j) which stands for the j'th input to the i'th module
+        in the graph.
+        """
+        m = len(self.modules_list)
+        self.output_users: List[List[Tuple[int, int, int]]] = [[] for _ in range(m)]
+        self.model_input_users = []
+        for i, input in enumerate(self.inputs):
+            assert input is not None
+            for j, input_item in enumerate(input):
+                if input_item[0] >= 0:
+                    self.output_users[input_item[0]].append((i, j, input_item[1]))
+                else:
+                    self.model_input_users.append((i, j, input_item[1]))
+
+
+class DistributedPipelineRecord:
+    """ A class for storing a single mini-batch (consisting of multiple micro-batches) as input to
+    a single partition.
+    Args:
+        device: the local device that runs the partition.
+        rank: the rank of the partition in the pipeline.
+        chunks: number of micro-batches in a mini-batch
+        num_inputs: number of inputs to the partition.
+        users: list of consumers of outputs of the partition. Each consumer in the list is a tuple
+            (remote_partition_rref, input_idx, output_idx) where remote_partition_rref points to a
+            remote DistributedPipelineRecord for consumer partiton for this mini-batch. The output number
+            output_idx of this partition will be used as the input number input_idx of that partition.
+    """
+
+    def __init__(
+        self, device: torch.device, rank: int, chunks: int, num_input: int, users: List[Tuple[rpc.RRef, int, int]]
+    ) -> None:
+        self.ready_cv = Condition()
+        # Each chunk consists of num_input tensors. self.tensors stores these individual tensors.
+        self.tensors: List[List[Optional[Tensor]]] = [[None] * num_input for _ in range(chunks)]
+        # For each tensor in self.tensors, we record a cuda event in corrsponding tensorpipe stream in self.recv_events,
+        # and later the stream that processes that tensor will wait on that event.
+        self.recv_events = [[None] * num_input for _ in range(chunks)]
+        # Once all num_input tensors of a given chunk are recieved, they are assembled as a batch and stored in
+        # self.batches
```

Comment:
redundant comment. maybe remove or replace?

Reply:
I feel it needs some comment here, to explain why we need both self.tensors and self.batches.